contracts/clients/src/BlsSigner.ts

```diff
@@ -0,0 +1,210 @@
+    this.#verifyInit();
+    const chainId = await this.getChainId();
+    if (!nonce) {
+      nonce = await BlsWalletWrapper.Nonce(
+        this.wallet.PublicKey(),
+        this.verificationGatewayAddress,
+        this.provider,
+      );
+    }
+
+    // TODO: bls-wallet #412 Update values returned in bundle receipt to more closely match ethers transaction response
+    return {
+      hash,
+      confirmations: 1,
+      from,
+      nonce: BigNumber.from(nonce).toNumber(),
+      gasLimit: BigNumber.from("0x0"),
+      value: BigNumber.from(action.ethValue),
+      data: action.encodedFunction.toString(),
+      chainId: chainId,
+      wait: (confirmations?: number) => {
+        return this.provider.waitForTransaction(hash, confirmations);
+      },
+    };
+  }
+
+  async unlock(password: string): Promise<boolean> {
+    const provider = this.provider;
+
+    const address = await this.getAddress();
+
+    return provider.send("personal_unlockAccount", [
+      address.toLowerCase(),
+      password,
+      null,
+    ]);
+  }
+
+  override async signTransaction(
+    transaction: Deferrable<ethers.providers.TransactionRequest>,
+  ): Promise<string> {
+    throw new Error(
+      "signTransaction() is not implemented, call 'signBlsTransaction()' instead.",
+    );
+  }
+
+  async signBlsTransaction(action: ActionData): Promise<Bundle> {
+    this.#verifyInit();
+    const nonce = await BlsWalletWrapper.Nonce(
+      this.wallet.PublicKey(),
+      this.verificationGatewayAddress,
+      this,
+    );
+
+    return this.wallet.sign({ nonce, actions: [action] });
+  }
+
+  override async signMessage(message: Bytes | string): Promise<string> {
+    throw new Error("signMessage() is not implemented.");
+  }
+
+  /** Sign a message */
+  signBlsMessage(message: string): Signature {
+    this.#verifyInit();
+    return this.wallet.signMessage(message);
+  }
+
+  override connect(provider: ethers.providers.Provider): BlsSigner {
+    throw new Error("connect() is not implemented.");
+  }
+
+  async _signTypedData(
+    domain: any,
+    types: Record<string, Array<any>>,
+    value: Record<string, any>,
+  ): Promise<string> {
+    throw new Error("_signTypedData() is not implemented.");
+  }
+
+  connectUnchecked(): ethers.providers.JsonRpcSigner {
+    throw new Error("connectUnchecked() is not implemented.");
+  }
+
+  async sendUncheckedTransaction(
+    transaction: Deferrable<ethers.providers.TransactionRequest>,
+  ): Promise<string> {
+    throw new Error("sendUncheckedTransaction() is not implemented.");
+  }
+
+  async _legacySignMessage(message: Bytes | string): Promise<string> {
+    throw new Error("_legacySignMessage() is not implemented.");
+  }
+
+  #verifyInit = () => {
+    if (!this.wallet || !this.verificationGatewayAddress) {
+      throw new Error(
+        "To perform this operation, ensure you have instantiated a BlsSigner and have called this.init() to initialize the wallet.",
+      );
+    }
+  };
+}
```

Comment:
Rather than doing this, you can store an init promise in the constructor and `await` its completion in async call points.

```typescript
class BlsSigner {
  readonly initPromise:Promise<void>; 
  
  constructor() {
    this.initPromise = this.init();
  };

  private async init(): Promise<void> {
    return Promise.resolve();
  }

  public async getNumber(): Promise<number> {
    await this.initPromise();
    return 42;
  }
}

const signer = new BlsSigner();
const num = await signer.getNumber();
```

Reply:
Updated with this solution